### PR TITLE
Should not coerce children prop on custom elements to a string.  Fixes #5088

### LINKED ISSUE #5088: Custom elements `children` prop gets coerced to string
Hello,

In React 0.13.3 HTML5 custom tags were rendered fine. For example:

```javascript
//... render function
  return (
    <my-custom-tag className="active custom-class" >
      <div>Hello World</div>
    </my-custom-tag>
  );
```

would render:
```html
<my-custom-tag class="active custom-class" data-reactid=".1">
  <div>Hello World</div>
</my-custom-tag>
```

Since I am in the process of migrating to v0.14.0 i saw that the above renders like this:
```html
<my-custom-tag classname="active custom-class" children="[object Object]" data-reactid=".1">
  <div>Hello World</div>
</my-custom-tag>
```

I can set in the JSX the property to be `class` instead of `className` and it will render with `class` attribute in the custom tag, but I am afraid this is not the intended behavior. If this is the intended behavior please close the issue and give me some documentation, because I tried to find something related to this for the last 5 hours.

### PR DESCRIPTION
**Pull request by jimfb**
Should not coerce children prop (or any prop, really) on custom elements to a string.  Fixes #5088

### CODE REVIEW

**Comment by jimfb** [comment]
@spicyj

**Comment by spicyj** [comment]
Should we just treat children (and dangerouslySetInnerHTML) specially? It seems maybe-useful to cast other things to string.

**Comment by jimfb** [comment]
Yeah, that's what my original diff (https://github.com/jimfb/react/commit/a703c3783b66675346855676680ef5604088248c) did, but then I thought about it and most of the coercions were mistakes.  There are situations where we might want to treat non-string values differently (for instance, a function property following the pattern onXXX might be treated as an event handler).  Figured we could always relax the coercion requirement in the future, but for now requiring it to be a string/number seems like a reasonable restriction.

**Comment by jimfb** [comment]
Ping @spicyj.  Have I convinced you that we should require all webcomponent props to be strings/numbers, and perhaps back off on this requirement later.  Or would you prefer to go with the older diff (https://github.com/jimfb/react/commit/a703c3783b66675346855676680ef5604088248c).  Or what's your thinking?

**Comment by sebmarkbage** [comment]
Regardless, this is a breaking change so it should not be a patch level release. It can go into 0.15.

**Comment by jimfb** [comment]
@sebmarkbage Addresses a regression that we introduced, namely that we now set the `children` attribute to `[Object object]` for all web components.

**Comment by sebmarkbage** [comment]
Then we should do what @spicyj suggested and your original.

JS has lots of coercion. It's just how the language works. Silently ignoring isn't necessarily better. It is probably worse here since it becomes a breaking change rather than an isolated patch of the regression.

If you want to protect against it, use [Flow](http://flowtype.org/) or something like [Restrict Mode](http://restrictmode.org/) on top.

**Comment by jimfb** [comment]
@spicyj @sebmarkbage PR updated; ready to merge?

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 1]
```diff
@@ -688,9 +688,11 @@ ReactDOMComponent.Mixin = {
           }
           propValue = CSSPropertyOperations.createMarkupForStyles(propValue);
         }
-        var markup = null;
+        var markup = '';
         if (this._tag != null && isCustomComponent(this._tag, props)) {
-          markup = DOMPropertyOperations.createMarkupForCustomAttribute(propKey, propValue);
+          if (propKey != 'children' || typeof propValue === 'string' || typeof propValue === 'number') {
```
Comment (sebmarkbage) [inline src/renderers/dom/shared/ReactDOMComponent.js]: Plain string comparison is not closure compiler compatible.

We should ignore it even if it's a string/number.

Otherwise this won't work:

```js
<custom-element>My Text Content</custom-element>
```

Comment (zpao) [inline src/renderers/dom/shared/ReactDOMComponent.js]: To get past the closure compiler issue you can use `var CHILDREN = keyOf({children: null})` and then use that in place of the string literal.

That's the safest thing, though `children` might actually be in the DOM externs.

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 2]
```diff
@@ -688,9 +689,11 @@ ReactDOMComponent.Mixin = {
           }
           propValue = CSSPropertyOperations.createMarkupForStyles(propValue);
         }
-        var markup = null;
+        var markup = '';
         if (this._tag != null && isCustomComponent(this._tag, props)) {
-          markup = DOMPropertyOperations.createMarkupForCustomAttribute(propKey, propValue);
+          if (propKey != CHILDREN || typeof propValue === 'string' || typeof propValue === 'number') {
```
Comment (zpao) [inline src/renderers/dom/shared/ReactDOMComponent.js]: `!==` (looks like we might have relaxed that when switching to fbjs lint rules)

@spicyj's long line length detector is slipping…

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 3]
```diff
@@ -970,6 +973,9 @@ ReactDOMComponent.Mixin = {
           deleteListener(this._rootNodeID, propKey);
         }
       } else if (isCustomComponent(this._tag, nextProps)) {
+        if (propKey == 'children' && typeof nextProp !== 'string' && typeof nextProp !== 'number') {
```
Comment (zpao) [inline src/renderers/dom/shared/ReactDOMComponent.js]: `===` and use `CHILDREN`

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 4]
```diff
@@ -49,6 +49,7 @@ var registrationNameModules = ReactBrowserEventEmitter.registrationNameModules;
 
 // For quickly matching children type, to test if can be treated as content.
 var CONTENT_TYPES = {'string': true, 'number': true};
+var CHILDREN = keyOf({children: null});
 
 var STYLE = keyOf({style: null});
```
Comment (zpao) [inline src/renderers/dom/shared/ReactDOMComponent.js]: nit: Want to put `CHILDREN` with `STYLE`. And actually if you rebase, you'll pick up the new one here also using keyof, would be good to have the group of them together

**Comment by zpao** [comment]
@jimfb I'm going to do another 0.14.2 soon so if you can fix up the last few things we can probably get this in there.

Although… I don't see history of when this got changed from 0.15 to 0.14.1, is it actually safe?

**Comment by jimfb** [comment]
Done.  I rebased, but don't see the new `keyof`, though that's probably not terribly important.

This was a regression introduced in 0.14, so it **should** be safe.

**Comment by zpao** [comment]
Well, this doesn't quite bring us back to the same behavior of 0.13. So regression… sort of yes. But the changes here don't "fix" everything. I don't know what's right.

I think there's also an argument to be made for explicit use of the `children` prop vs the rest arg use. The rest arg use has clear intention of putting them as actual child nodes. The `children` prop is less clear (and unfortunately a big cloudy when we consider composition and `{...this.props}`).

```js
React.renderToStaticMarkup(<my-component>5</my-component>)
React.renderToStaticMarkup(<my-component children={5}></my-component>)
React.renderToStaticMarkup(<my-component>{[5]}</my-component>)

// 0.13.3
'<my-component>5</my-component>'
'<my-component>5</my-component>'
'<my-component>5</my-component>'

0.14.1
'<my-component children="5">5</my-component>'
'<my-component children="5">5</my-component>'
'<my-component children="5">5</my-component>'

// With this
'<my-component children="5">5</my-component>'
'<my-component children="5">5</my-component>'
'<my-component>5</my-component>'
```

As you can see, you've only changed that last one. Which is consistent with the "not coercing children" issue as filed. Maybe that's correct and I should shut up, it just seems weird.

**Comment by jimfb** [comment]
@zpao So the "correct" behavior, imo, is to have `<my-component children="5" />` != `<my-component>5</my-component>` at the element level.  It's fine if you want to merge children into props.children before passing them into composite components in order to preserve API, but at the element level they should be separate.

Sebastian wants (at least for now) us to drop the typeof checks, so I'll do that now.

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 5]
```diff
@@ -742,9 +743,11 @@ ReactDOMComponent.Mixin = {
           }
           propValue = CSSPropertyOperations.createMarkupForStyles(propValue, this);
         }
-        var markup = null;
+        var markup = '';
```
Comment (sebmarkbage) [inline src/renderers/dom/shared/ReactDOMComponent.js]: Why is this change needed?

Comment (jimfb) [inline src/renderers/dom/shared/ReactDOMComponent.js]: No idea.  Presumably I wouldn't have made it if it wasn't necessary, but I don't see the reason for it and it passes unit tests as null so I'll go ahead and remove it.

**Comment by sebmarkbage** [comment]
I think that it is likely that if we wanted to "standardize" the outline of a React Element, that other people might have opinion on this part of the structure. If it becomes necessary, I'd like to punt changing the React Element outline until that point.
